specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1256 @@
+
+Windows provides access to a process' package graph primarily via [GetCurrentPackageInfo](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentpackageinfo) and [GetCurrentPackageInfo2](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentpackageinfo2). Equivalent information is available via [Windows.ApplicationModel.Package.Current](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.package.current?view=winrt-19041) and [Package.Current.Dependencies](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.package.dependencies?view=winrt-19041). These correspond to PackageGraph[0] and PackageGraph[1+]. These APIs provide access to the static package graph.
+
+This information is used by several Windows components to find resources across packages available to a process (aka across a process' package graph) including:
+
+- **DLLs** - the Loader searches the package graph for DLLs (per [Dynamic-Link Library Search Order](https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order))
+- **CLR** - the CLR searches the package graph for assemblies (similar to how the Loader does so for DLLs)
+- **WinRT** - WinRT searches the package graph for WinRT activation. Type resolution is constrained to types provided by packages in a process' package graph (and those provided by Windows via the `Windows.*` namespace). Thius include metadata resolution (i.e. *.winmd).
+- **MRT** - MRT searches the package graph for resources (resources.pri, images, etc)
+- **ms-appx URI scheme** - [ms-appx URIs](https://docs.microsoft.com/en-us/windows/uwp/app-resources/uri-schemes) are resolved to packages across the package graph
+
+If we can alter the package graph as seen by `GetCurrentPackageInfo` and like APIs we can provide the desired dynamic behavior.
+
+**Solution:** We'll manage our own package graph to supplement the static package graph (if any) with dynamic dependencies. We'll use [Detours](https://github.com/Microsoft/Detours) to hook `GetCurrentPackageInfo` and related APIs to redirect to our own Dynamic Dependencies savvy variants.
+
+### 3.1.1. Detours to Enhance Package Graph APIs
+
+Package Graph information is available for packages registered to the current user via [OpenPackageInfoByFullName](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-openpackageinfobyfullname). We'll keep track of the expanded package graph (static + dynamic) via a global variable filled with PACKAGE_INFO_REFRENCE objects. Our Detour'd API variants walk this information to produce their answers seamlessly spanning static and dynamic package information for the process.
+
+```c++ (C++ish pseudocode)
+// An instance with no PACKAGE_INFO_REFERENCE (nullptr) is a placeholder for the static package graph
+class PackageGraphNode
+{
+public:
+    PackageGraphNode() = default
```

Comment:
Is this spec-worthy or implementation detail?

All C++ snippets are actually C++ish pseudocode thus no copy ctor, no semicolons, no #includes, etc. Does it need to be (actual) C++ instead of pseudocode? Does it need to be rigorously correct C++ e.g. it must compile? Not a lot of standard practices to follow yet so I aimed for meaningfully informative w/o burning implementation time to spec not significant fine print.

The markdown has language annotations of `c++ (C++ish pseudocode)`, `c# (but really MIDL3)` and `xml` but I see none of that comes through in rendered HTML. Add a 1st line comment making that clearer?

Reply:
Maybe instead remove this and other type prototypes and just declare "we'll specify a C++ helper type that manages the lifecycle of a package graph node." ?

Reply:
I've gutted PackageGraphNode as it was mostly implementation detail that obviously wasn't helping provide more clarity. Kept a small definition with the data it manages as the key explanation what we're managing. Anywhere pseudocode has "..." that stands for "insert messy implementation details"